README.md
The Haskelling Homeschooler
===========================

A set of mathematical exercises for children, performed by 
implementing solutions in Haskell.

* [Exercise 1](exercise-1.md)
* [Exercise 2](exercise-2.md)
* [Exercise 3](exercise-3.md)
* [Exercise 4](src/Magic.hs)




ChangeLog.md
# Revision history for hs-curriculum

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Ex4.hs
module Ex4 where

data Color 
  = White
  | Blue
  | Black
  | Red
  | Green
  | Colorless
  deriving (Eq, Show)

readColor :: String -> Maybe Color
readColor "white" = Just White
readColor "blue"  = Just Blue
readColor "black" = Just Black
readColor "red"   = Just Red
readColor "green" = Just Green
readColor "none"  = Just Colorless
readColor _ = Nothing

data ColorCount = CC Color Int

main :: IO ()
main = do
  l <- promptLands
  r <- promptColor Red
  g <- promptColor Green
  print (Red,   landProp <$> r <*> g <*> l)
  print (Green, landProp <$> g <*> r <*> l)

landProp :: (Fractional n) => n -> n -> n -> n
landProp primary other total = 
  let frac = primary / (primary + other)
  in  frac * total

promptColor :: (Read n) => Color -> IO (Maybe n)
promptColor c = do
  prompt (concat ["Enter the number of ", show c, " color symbols: "])

promptLands :: (Read n) => IO (Maybe n)
promptLands = do
  prompt "Enter the number of lands: "

prompt :: (Read n) => String -> IO (Maybe n)
prompt p = do
  putStr p
  s <- getLine
  return (safeRead s)

safeRead :: (Read n) => String -> Maybe n
safeRead s = case reads s of 
  [] -> Nothing
  [(f, _)] -> Just f


src/Magic.hs
{-# LANGUAGE ScopedTypeVariables #-}
module Magic where

{- Create a data type whose constructors represent the different colors
 - of mana. `Color` is the *type* that we are defining, and `White`, 
 - `Blue`, `Black` etc. are each 0-argument functions that construct different
 - values of the type `Color`
 -}
data Color 
  = White
  | Blue
  | Black
  | Red
  | Green
  | Colorless
  deriving (Eq, Show)

{- `main` is the primary entry point for this program. To run the program, first
 - open the Haskell interpreter by typing `stack ghci` at the terminal prompt
 - from the directory containing this file. Then, when you see the `Prelude>`
 - prompt, you can type:
 -
 - Prelude> :l Magic.hs
 -
 - to load this file. Once the file has loaded (which includes compilation; if
 - there are any errors, they will appear here) you can type:
 -
 - *Magic> main
 -
 - to run the program.
 -}
main :: IO ()
main = do
  l <- promptLandCount
  r <- promptSymbolCount Red
  g <- promptSymbolCount Green
  putStrLn (concat ["Use ", show (landFrac <$> r <*> g <*> l), " ", landName Red])
  putStrLn (concat ["Use ", show (landFrac <$> g <*> r <*> l), " ", landName Green])

{- Prompt the user for total number of lands to include in their deck. In
 - a 60-card deck, 24 is a good starting point for this value; in a 40-card
 - deck, try 17.k
 -}
promptLandCount :: IO (Maybe Int)
promptLandCount = do
  prompt "Enter the total number of lands you want to have in your deck: "

{- Prompt the user for the count of symbols for the specified color.
 -}
promptSymbolCount :: Color -> IO (Maybe Int)
promptSymbolCount c = do
  prompt (concat ["Enter the number of ", show c, " color symbols: "])

{- This function performs the central arithmetic logic of the program
 - by determining what percentage of the overall land population needs
 - to be of the 'primary' color, and then multiplying that percentage
 - by the total desired number of lands.
 -}
landFrac :: Int -- the count of color symbols for the color for which you want a land count
         -> Int -- the total count of color symbols which are *not* the primary color
         -> Int -- the total number of lands desired for the final deck
         -> Int -- the result will be the number of lands of the primary color that you should use
landFrac primary other lands = 
  let primary' :: Double -- convert the number of primary-color symbols to a double-precision floating point value, for division
      primary' = fromIntegral primary 
      
      other' :: Double -- convert the number of other mana symbols to a double-precision floating point value, for division
      other' = fromIntegral other

      frac :: Double -- the overall fraction of primary lands is a double-precision floating point value
      frac = primary' / (primary' + other')

  in  round (frac * fromIntegral lands) -- round the result to an integer value

{- A convenience function for displaying the basic land type for a each color.
 - Note that there is one implementation of this function for each color's
 - constructor - this is called "pattern matching"
 -}
landName :: Color -> String
landName White = "plains"
landName Blue = "islands"
landName Black = "swamps"
landName Red = "mountains"
landName Green = "forests"
landName Colorless = "wastes"

{- For future use: this function will "parse" a string to turn it into
 - a color value. Use this if you want to add functionality to prompt
 - the user for what color they want to enter.
 -}
readColor :: String -> Maybe Color
readColor "white" = Just White
readColor "blue"  = Just Blue
readColor "black" = Just Black
readColor "red"   = Just Red
readColor "green" = Just Green
readColor "none"  = Just Colorless
readColor _ = Nothing

{- A utility function for prompting the user for a value. This eliminates
 - the need for us to duplicate code for different prompts.
 -}
prompt :: (Read n) => String -> IO (Maybe n)
prompt p = do
  putStr p
  s <- getLine
  return (safeRead s)

{- This function is a helper to ensure that if the user enters a bad
 - value, the program terminates cleanly rather than failing with an
 - error that may be difficult to understand. We will improve on this
 - function in a future version.
 -}
safeRead :: (Read n) => String -> Maybe n
safeRead s = case reads s of 
  [] -> Nothing
  [(f, _)] -> Just f


src/LoH/Game.hs
module LoH.Game where

import           Introlude
import qualified Data.Map.Strict  as M

newtype Location where
  ToLoc :: { locName :: Text } -> Location
  deriving (Eq, Ord)

livingRoom :: Location
livingRoom = ToLoc { locName = "living-room" }

garden = ToLoc { locName = "garden" }
garden = undefined

attic = ToLoc { locName = "attic" }
attic = undefined

newtype Description where
  ToDesc :: { descText :: Text } -> Description

type Locations = M.Map Location Description

locations :: Locations
locations =
  M.fromList [ (livingRoom, ToDesc "You are in the living-room. A wizard is snoring loudly on the couch.")
             , (garden, ToDesc "You are in a beautiful garden. There is a well in front of you.")
             , (attic, ToDesc "You are in the attic. There is a giant welding torch in the corner.")
             ]

describeLocation :: Locations -> Location -> Maybe Description
describeLocation locs loc = M.lookup loc locs

data Path where
  ToPath :: { dest :: Location, dir :: Text, via :: Text } -> Path

type Paths = M.Map Location [Path]

paths :: Paths
paths =
  M.fromList [ (livingRoom, [ ToPath { dest = garden, dir = "west", via = "door" }
                            , ToPath { dest = attic,  dir = "upstairs", via = "ladder" }])
             , (garden,     [ ToPath { dest = livingRoom, dir = "east", via = "door" }])
             , (attic,      [ ToPath { dest = livingRoom, dir = "downstairs", via = "ladder" }])
             ]

describePath :: Path -> Text
describePath p =
  concat ["There is a ", via p," going ", dir p, " from here"]

describePathsFrom :: Paths -> Location -> [Text]
describePathsFrom ps loc =
  maybe
    ["I can't find that location!"]
    (\locPaths -> mapList describePath locPaths)
    (M.lookup loc ps)

newtype Item where
  ToItem :: { itemName :: Text } -> Item
  deriving (Eq, Ord)

type Items = M.Map Location [Item]

items :: Items
items =
  M.fromList [ (garden, mapList ToItem ["frog", "chain"])
             , (livingRoom, mapList ToItem ["whiskey", "bucket"])
             ]

data Environment where
  Env :: { currentLocation :: Location
         , envLocations :: Locations
         , envPaths :: Paths
         , envItems :: Items
         } -> Environment


src/Introlude.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

-- | An extremely minimal prelude for children learning Haskell.
--
-- This prelude provides monomorphized versions of a number of
-- commonly used polymorphic functions. This is intended to give
-- the learner a very concrete environment in which to become familiar
-- with the language.
module Introlude
  ( -- IO
    Action
  , putTextAction
  , getTextAction
--  , toAction
  , mapAction
  , composeActionFns
  , withActionResult
  , andThen
  , runAction
  -- comparisons
  , P.Eq
  , P.Ord
  -- Ints
  , Int
  , addInt
  , divInt
  , renderInt
  -- Text
  , Text
  , T.concat
  , T.intercalate
  -- List
  , toList
  , mapList
  , withListElements
  -- Maybe
  , P.Maybe(..)
  , toMaybe
  , P.fromMaybe
  , P.maybe
  , P.undefined
  ) where

import qualified Prelude as P
import           Prelude (IO, Int, (.))

import qualified Control.Monad as P

import qualified Data.Maybe as P
import qualified Data.Text as T
import           Data.Text (Text)

-- | An Action represents a restricted I/O operation. For introductory
-- purposes, the only I/O operations supported are reading lines from
-- stdin and writing lines to stdout.
newtype Action a = Action { runAction :: IO a }
  deriving (P.Functor, P.Applicative, P.Monad)

instance P.Show (Action a) where
  show _ = "You constructed an Action value. You have to use the 'runAction' \
           \function to actually run it in the REPL to see what it does."

--toAction :: a -> Action a
--toAction = P.pure

-- | Construct an Action that, when you run it, will print the provided
-- line of text to the screen.
putTextAction :: Text -> Action ()
putTextAction t = Action (P.putStrLn (T.unpack t))

-- | Construct an Action that, when you run it, will read one line of
-- input from the console. The program will pause while you type in text;
-- when you hit \<Enter\> it will continue. You can use 'composeActionFns'
-- or 'withActionResult' to actually use the text that was read in inside
-- your program.
getTextAction :: Action Text
getTextAction = Action (P.fmap T.pack P.getLine)

-- | Create a new action that first performs the first action, then performs
-- the second when run. Most programs need to do more than one thing that
-- requires input or output, so this is a way to join together multiple
-- actions so that they both happen when the resulting /composite/ 'Action' is run.
andThen :: Action () -> Action b -> Action b
andThen = (P.>>)

mapAction :: (a -> b) -> Action a -> Action b
mapAction = P.fmap

composeActionFns :: (a -> Action b) -> (b -> Action c) -> (a -> Action c)
composeActionFns = (P.>=>)

withActionResult :: Action a -> (a -> Action b) -> Action b
withActionResult = (P.>>=)

-------

addInt :: Int -> Int -> Int
addInt = (P.+)

divInt :: Int -> Int -> Int
divInt = P.div

renderInt :: Int -> Text
renderInt = T.pack . P.show

toList :: a -> [a]
toList = P.pure

mapList :: (a -> b) -> [a] -> [b]
mapList = P.fmap

withListElements :: (a -> [b]) -> [a] -> [b]
withListElements = (P.=<<)

toMaybe :: a -> P.Maybe a
toMaybe = P.pure


--{-# INLINE toAction #-}
{-# INLINE putTextAction #-}
{-# INLINE getTextAction #-}
{-# INLINE andThen #-}
{-# INLINE withActionResult #-}
{-# INLINE mapAction #-}
{-# INLINE addInt #-}
{-# INLINE divInt #-}
{-# INLINE renderInt #-}
{-# INLINE toList #-}
{-# INLINE mapList #-}
{-# INLINE withListElements #-}
{-# INLINE toMaybe #-}


src/LoH/GuessMe.hs
module LoH.GuessMe where

import Introlude

-- | Implement a function that returns a guess that is half way between the two
-- arguments
guess :: Int -> Int -> Int
guess = undefined

-- | Given a number <n>, return an Action that will print out the text "My
-- guess is <n>." For example, `printGuess 10` will return an Action that when
-- run causes "My guess is 10" to be printed to the screen.
printGuess :: Int -> Action ()
printGuess = undefined

-- | This is the main game. It should be implemented as a recursive function
-- that takes as its arguments the lower and upper bounds of the number to be
-- guessed.
--
-- First, it should compute a guess using the 'guess' function , and
-- display it to the player using `'printGuess'.
--
-- Then, it should accept a line of input. If the input is "smaller" then it
-- should call itself recursively, setting the upper bound to the guess; if the
-- input is "bigger" then it should make the recursive call using the guess for
-- the lower bound. It should also accept the "quit" string, in which case it
-- prints "Goodbye!" and returns without calling itself; if any other input is
-- received, it should print an error and repeat the request for input.
guessMyNumber :: Int -> Int -> Action ()
guessMyNumber = undefined

-- | This is the main function used to start the game. It has already been
-- implemented for you; it demonstrates how to use the `andThen` function to
-- join together two actions.
startGame =
  runAction (
    (putTextAction "A number guessing game??? Awesome!")
    `andThen`
    (guessMyNumber 0 100)
  )



src/Magic/LandCalc.hs
{-# LANGUAGE ScopedTypeVariables #-}
module Magic.LandCalc where

import Prelude

{- Create a data type whose constructors represent the different colors
 - of mana. `Color` is the *type* that we are defining, and `White`,
 - `Blue`, `Black` etc. are each 0-argument functions that construct different
 - values of the type `Color`
 -}
data Color
  = White
  | Blue
  | Black
  | Red
  | Green
  | Colorless
  deriving (Eq, Show)

{- `main` is the primary entry point for this program. To run the program, first
 - open the Haskell interpreter by typing `stack ghci` at the terminal prompt
 - from the directory containing this file. Then, when you see the `Prelude>`
 - prompt, you can type:
 -
 - Prelude> :l Magic.hs
 -
 - to load this file. Once the file has loaded (which includes compilation; if
 - there are any errors, they will appear here) you can type:
 -
 - *Magic> main
 -
 - to run the program.
 -}
main :: IO ()
main = do
  l <- promptLandCount
  r <- promptSymbolCount Red
  g <- promptSymbolCount Green
  putStrLn (concat ["Use ", show (landFrac <$> r <*> g <*> l), " ", landName Red])
  putStrLn (concat ["Use ", show (landFrac <$> g <*> r <*> l), " ", landName Green])

{- Prompt the user for total number of lands to include in their deck. In
 - a 60-card deck, 24 is a good starting point for this value; in a 40-card
 - deck, try 17.k
 -}
promptLandCount :: IO (Maybe Int)
promptLandCount = do
  prompt "Enter the total number of lands you want to have in your deck: "

{- Prompt the user for the count of symbols for the specified color.
 -}
promptSymbolCount :: Color -> IO (Maybe Int)
promptSymbolCount c = do
  prompt (concat ["Enter the number of ", show c, " color symbols: "])

{- This function performs the central arithmetic logic of the program
 - by determining what percentage of the overall land population needs
 - to be of the 'primary' color, and then multiplying that percentage
 - by the total desired number of lands.
 -}
landFrac :: Int -- the count of color symbols for the color for which you want a land count
         -> Int -- the total count of color symbols which are *not* the primary color
         -> Int -- the total number of lands desired for the final deck
         -> Int -- the result will be the number of lands of the primary color that you should use
landFrac primary other lands =
  let primary' :: Double -- convert the number of primary-color symbols to a double-precision floating point value, for division
      primary' = fromIntegral primary

      other' :: Double -- convert the number of other mana symbols to a double-precision floating point value, for division
      other' = fromIntegral other

      frac :: Double -- the overall fraction of primary lands is a double-precision floating point value
      frac = primary' / (primary' + other')

  in  round (frac * fromIntegral lands) -- round the result to an integer value

{- A convenience function for displaying the basic land type for a each color.
 - Note that there is one implementation of this function for each color's
 - constructor - this is called "pattern matching"
 -}
landName :: Color -> String
landName White = "plains"
landName Blue = "islands"
landName Black = "swamps"
landName Red = "mountains"
landName Green = "forests"
landName Colorless = "wastes"

{- For future use: this function will "parse" a string to turn it into
 - a color value. Use this if you want to add functionality to prompt
 - the user for what color they want to enter.
 -}
readColor :: String -> Maybe Color
readColor "white" = Just White
readColor "blue"  = Just Blue
readColor "black" = Just Black
readColor "red"   = Just Red
readColor "green" = Just Green
readColor "none"  = Just Colorless
readColor _ = Nothing

{- A utility function for prompting the user for a value. This eliminates
 - the need for us to duplicate code for different prompts.
 -}
prompt :: (Read n) => String -> IO (Maybe n)
prompt p = do
  putStr p
  s <- getLine
  return (safeRead s)

{- This function is a helper to ensure that if the user enters a bad
 - value, the program terminates cleanly rather than failing with an
 - error that may be difficult to understand. We will improve on this
 - function in a future version.
 -}
safeRead :: (Read n) => String -> Maybe n
safeRead s = case reads s of
  [] -> Nothing
  [(f, _)] -> Just f
